### DISCUSSION TITLE
Proximity between training and testing objects using a multivariate forest

### DISCUSSION THREAD

**Discussion by lennartnoordermeer** [kogalur/randomForestSRC#360] [Q&A] [answered]
I trained a multivariate forest:

      # example data:
      library("randomForestSRC")
      data(nutrigenomic)
      ydta <- nutrigenomic$lipids
      xdta <- data.frame(nutrigenomic$genes,
                         diet = nutrigenomic$diet,
                         genotype = nutrigenomic$genotype)
      # multivariate forest call:
      mv.obj <- rfsrc(get.mv.formula(colnames(ydta)),
                      data.frame(ydta, xdta),
                      importance=TRUE, nsplit = 10)
When I want to predict the (multivariate) outcomes for new objects, I use:

      # select first 10 rows as test data:
      testdta <- data.frame(nutrigenomic$genes+0.01,
                         diet = nutrigenomic$diet,
                         genotype = nutrigenomic$genotype)[1:10,]
      # predict:
      pred <- predict(object=mv.obj, newdata=testdta, proximity = T)
      
I can extract a proximity matrix:

      prox <- pred$proximity
However this gives only the proxmities among the new objects (testdta). Can I calculate the proximities between the new objects (in testdta) and the training set (xdta) based on the already existing forest (mv.obj)? So the proximity matrix I am looking for has the training objects as rows (n=40) and new objects as columns (n=10), or vice versa. 

I could run rfsrc() unsupervised again on a combined data set to obtain the proximities (a 50 x 50 matrix):

      # unsupervised forest:
      unspv <- rfsrc(data = rbind(xdta,testdta))
      prox <- predict(object=unspv,newdata=LargeData, proximity = T)$proximity

However I think there must be some way to avoid building the forest again and instead using the already existing one.

Thank you! Lennart

**Comment by ishwaran** [comment, marked as answer]
To obtain proximity values for both test, train:

`> pred <- predict(object=mv.obj, newdata=rbind(testdta, xdta), proximity = T)`

Notice the dimension is 50x50 

```
> dim(pred$proximity)
[1] 50 50
```

The upper diagonal matrix is the test data, the lower diagonal is the training data, and the off diagonals are the proximities for (test, train) and (train, test), respectively.

**Comment by ishwaran** [comment]
Also, if you wanted just the proximity of the training data, and you are interested in OOB values for this (or inbag, etc.) then you could use the following 

`predict(mv.obj, proximity = "oob")$proximity`

`predict.rfsrc` can be used on the grow call to extract all kinds of quantities.  This makes it possible to later obtain quantities like importance, proximity, error rates and so forth that might have slowed the original forest building.

See the help file for [predict.rfsrc](https://www.randomforestsrc.org/reference/predict.rfsrc.html) for further details.
